[AT-5184] Retry failed portfolio provisioning jobs

graham-dds opened the pull request:
Closes: [AT-5184](https://ccpo.atlassian.net/browse/AT-5184)

This PR ensures that CSP exceptions raised during provisioning get bubbled up so that Celery can retry them.

When our `provision_portfolio` Celery task is called, we move the state machine of a given portfolio to an `*_IN_PROGRESS` state and call `after_in_progress_callback`. All that's needed for Celery to retry the task is for `after_in_progress_callback` to raise a `GeneralCSPException`. This now happens in `PortfolioStateMachine._do_provision_stage`.

At present, if a portfolio fails a provisioning step after 3 retries (the Celery default), it will remain in a failed state and not get picked up again for provisioning.

dandds commented on atat/models/portfolio_state_machine.py:
```diff
@@ -123,6 +123,19 @@ def current_state(self):
+        """
+        for stage_state in StageStates:
+            if self.current_state.name.endswith(stage_state.name):
+                stage, stage_state = self.current_state.name.split(
```
Nitpicky, but maybe `_stage_state` since it's unused.

graham-dds replied:
It's used on 133 --  `...endswith(stage_state.name)`

dandds replied:
aaah missed that, sorry

graham-dds commented:
@dandds Ripped out all of the logging and granular exception handling.

I couldn't figure out how to test, `"General CSP exceptions trigger retries, but other exceptions do not`, though I was able to confirm the expected behavior locally. It seemed like `provision_portfolio.retry` was always being called, but maybe I was setting up the mocking wrong.

graham-dds commented on atat/models/mixins/state_machines.py:
```diff
@@ -175,22 +176,40 @@ class FSMMixin:
             app.logger.info(
-                f"could not locate fail trigger '{fail_trigger}' for '{self.__repr__()}'"
+                f"could not locate {trigger_prefix} trigger '{trigger}' for '{self.__repr__()}'"
             )
```
Fine with me!

graham-dds replied:
https://github.com/dod-ccpo/atst/pull/1497/commits/01486a302bac409e34a562cc617b59deac088477

graham-dds commented on atat/jobs.py:
```diff
@@ -251,9 +252,9 @@ def do_provision_portfolio(csp: CloudProviderInterface, portfolio_id=None):
+@celery.task(bind=True, base=RecordFailure, autoretry_for=(GeneralCSPException,))
 def provision_portfolio(self, portfolio_id=None):
-    do_work(do_provision_portfolio, self, app.csp.cloud, portfolio_id=portfolio_id)
+    do_provision_portfolio(app.csp.cloud, portfolio_id=portfolio_id)
```
https://ccpo.atlassian.net/browse/AT-5217

graham-dds commented on atat/models/portfolio_state_machine.py:
```diff
@@ -123,147 +121,62 @@ def current_state(self):
+            if "complete" in self.machine.get_triggers(state_obj.name):
                 app.logger.info(
                     "last stage completed. transitioning to COMPLETED state"
                 )
```
https://github.com/dod-ccpo/atst/pull/1497/commits/01486a302bac409e34a562cc617b59deac088477

graham-dds commented:
@dandds I think I addressed all of your comments in my last commit (https://github.com/dod-ccpo/atst/pull/1497/commits/01486a302bac409e34a562cc617b59deac088477)
